File: csharp/Workers/PaymentCompletionWorker.cs
using Camunda.Orchestration.Sdk;

namespace Camunda.Training.CSharp.Workers;

public class PaymentCompletionWorker(CamundaClient client) : Worker("payment-completion", client)
{
    public override async Task<object?> Handler(ActivatedJob job, CancellationToken ct)
    {
        Console.WriteLine("Handling payment completion");

        var variables = job.GetVariables<Dictionary<string, object>>();
        if (variables != null
            && variables.TryGetValue("orderId", out object? orderIdObj)
            && orderIdObj?.ToString() is string orderId)
        {
            // Notify the waiting Order Process that payment is complete.
            // No need to send variables back; only the orderId correlation is needed.
            await client.PublishMessageAsync(new MessagePublicationRequest
            {
                Name = "paymentCompletedMessage",
                CorrelationKey = orderId,
                TimeToLive = (long)TimeSpan.FromMinutes(5).TotalMilliseconds,
            }, ct);

            Console.WriteLine("Message paymentCompletedMessage sent");
        }
        else
        {
            Console.WriteLine("orderId not found or not a string in the variables");
        }

        // Returning null auto-completes the job with no variables.
        return null;
    }
}

File: java-spring/src/main/java/io/camunda/training/workers/PaymentCompletionWorker.java
package io.camunda.training.workers;

import io.camunda.client.CamundaClient;
import io.camunda.client.annotation.JobWorker;
import io.camunda.client.annotation.Variable;
import io.camunda.client.api.response.ActivatedJob;
import io.camunda.client.api.worker.JobClient;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

@Component
public class PaymentCompletionWorker {

  Logger logger = LoggerFactory.getLogger(PaymentCompletionWorker.class);

  private final CamundaClient camundaClient;

  @Autowired
  public PaymentCompletionWorker(CamundaClient camundaClient) {
    this.camundaClient = camundaClient;
  }

  @JobWorker(type = "payment-completion", autoComplete = false)
  public void handlePaymentCompletion(JobClient jobClient, ActivatedJob job,
                                      @Variable String orderId) {
    logger.info("Job handled: {}", job.getType());

    // Tell Order Process to continue
    camundaClient.newPublishMessageCommand()
            .messageName("paymentCompletedMessage")
            .correlationKey(orderId)
            .send().join();

    // Complete the job in the Payment Process
    jobClient.newCompleteCommand(job).send();
  }
}